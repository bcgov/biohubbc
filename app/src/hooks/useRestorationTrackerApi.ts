import axios from 'axios';
import useAdminApi from './api/useAdminApi';
import useAxios from './api/useAxios';
import useCodesApi from './api/useCodesApi';
import useDraftApi from './api/useDraftApi';
import useExternalApi from './api/useExternalApi';
import useProjectApi, { usePublicProjectApi } from './api/useProjectApi';
import useSearchApi, { usePublicSearchApi } from './api/useSearchApi';
import useUserApi from './api/useUserApi';
import usePermitApi from './api/usePermitApi';
import { useContext } from 'react';
import { ConfigContext } from 'contexts/configContext';

/**
 * Returns a set of supported api methods.
 *
 * @return {*} object whose properties are supported api methods.
 */
export const useRestorationTrackerApi = () => {
  const config = useContext(ConfigContext);
  const apiAxios = useAxios(config?.API_HOST);

  const project = useProjectApi(apiAxios);

  const permit = usePermitApi(apiAxios);

  const search = useSearchApi(apiAxios);

  const codes = useCodesApi(apiAxios);

  const draft = useDraftApi(apiAxios);

  const user = useUserApi(apiAxios);

  const admin = useAdminApi(apiAxios);

  const external = useExternalApi(axios);

  const publicApis = {
    project: usePublicProjectApi(apiAxios),
    search: usePublicSearchApi(apiAxios)
  };

  return {
    project,
    permit,
    search,
    codes,
    draft,
    user,
    admin,
    external,
    public: publicApis
  };
};
