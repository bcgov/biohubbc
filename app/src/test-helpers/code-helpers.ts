import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';

export const codes: IGetAllCodeSetsResponse = {
  coordinator_agency: [{ id: 1, name: 'A Rocha Canada' }],
  management_action_type: [{ id: 1, name: 'Management action' }],
  first_nations: [{ id: 1, name: 'First nations code' }],
  funding_source: [{ id: 1, name: 'Funding source code' }],
  investment_action_category: [{ id: 1, fs_id: 1, name: 'Investment action category' }],
  activity: [{ id: 1, name: 'Activity code' }],
  project_type: [{ id: 1, name: 'Project type' }],
  region: [{ id: 1, name: 'Region code' }],
  species: [{ id: 1, name: 'Species code' }],
  proprietor_type: [
    { id: 1, name: 'Proprietor code 1', is_first_nation: false },
    { id: 2, name: 'First Nations Land', is_first_nation: true }
  ],
  iucn_conservation_action_level_1_classification: [{ id: 1, name: 'IUCN class' }],
  iucn_conservation_action_level_2_subclassification: [{ id: 1, iucn1_id: 1, name: 'IUCN subclass 1' }],
  iucn_conservation_action_level_3_subclassification: [{ id: 1, iucn2_id: 1, name: 'IUCN subclass 2' }],
  system_roles: [
    { id: 1, name: 'Role 1' },
    { id: 2, name: 'Role 2' }
  ],
  project_roles: [
    { id: 1, name: 'Project Role 1' },
    { id: 2, name: 'Project Role 2' }
  ],
  administrative_activity_status_type: [
    { id: 1, name: 'Pending' },
    { id: 2, name: 'Actioned' },
    { id: 3, name: 'Rejected' }
  ],
  common_survey_methodologies: [
    { id: 1, name: 'Recruitment' },
    { id: 2, name: 'SRB' }
  ]
};
