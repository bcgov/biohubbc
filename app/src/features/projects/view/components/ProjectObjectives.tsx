import Box from '@material-ui/core/Box';
import Divider from '@material-ui/core/Divider';
import { mdiPencilOutline } from '@mdi/js';
import Icon from '@mdi/react';
import EditDialog from 'components/dialog/EditDialog';
import { IErrorDialogProps } from 'components/dialog/ErrorDialog';
import ReadMoreField from 'components/fields/ReadMoreField';
import { H3ButtonToolbar } from 'components/toolbar/ActionToolbars';
import { EditObjectivesI18N } from 'constants/i18n';
import { DialogContext } from 'contexts/dialogContext';
import {
  IProjectObjectivesForm,
  ProjectObjectivesFormInitialValues,
  ProjectObjectivesFormYupSchema
} from 'features/projects/components/ProjectObjectivesForm';
import { APIError } from 'hooks/api/useAxios';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import {
  IGetProjectForUpdateResponseObjectives,
  IGetProjectForViewResponse,
  UPDATE_GET_ENTITIES
} from 'interfaces/useProjectApi.interface';
import React, { useContext, useState } from 'react';
import ProjectStepComponents from 'utils/ProjectStepComponents';

export interface IProjectObjectivesProps {
  projectForViewData: IGetProjectForViewResponse;
  codes: IGetAllCodeSetsResponse;
  refresh: () => void;
}

/**
 * Project objectives content for a project.
 *
 * @return {*}
 */
const ProjectObjectives: React.FC<IProjectObjectivesProps> = (props) => {
  const {
    projectForViewData: { objectives, id },
    codes
  } = props;

  const restorationTrackerApi = useRestorationTrackerApi();

  const dialogContext = useContext(DialogContext);

  const defaultErrorDialogProps = {
    dialogTitle: EditObjectivesI18N.editErrorTitle,
    dialogText: EditObjectivesI18N.editErrorText,
    open: false,
    onClose: () => {
      dialogContext.setErrorDialog({ open: false });
    },
    onOk: () => {
      dialogContext.setErrorDialog({ open: false });
    }
  };

  const showErrorDialog = (textDialogProps?: Partial<IErrorDialogProps>) => {
    dialogContext.setErrorDialog({ ...defaultErrorDialogProps, ...textDialogProps, open: true });
  };

  const [openEditDialog, setOpenEditDialog] = useState(false);

  const [objectivesDataForUpdate, setObjectivesDataForUpdate] = useState<IGetProjectForUpdateResponseObjectives>(
    null as any
  );
  const [objectivesFormData, setObjectivesFormData] = useState<IProjectObjectivesForm>(
    ProjectObjectivesFormInitialValues
  );

  const handleDialogEditOpen = async () => {
    let objectivesResponseData;

    try {
      const response = await restorationTrackerApi.project.getProjectForUpdate(id, [UPDATE_GET_ENTITIES.objectives]);

      if (!response?.objectives) {
        showErrorDialog({ open: true });
        return;
      }

      objectivesResponseData = response.objectives;
    } catch (error) {
      const apiError = error as APIError;
      showErrorDialog({ dialogText: apiError.message, open: true });
      return;
    }

    setObjectivesDataForUpdate(objectivesResponseData);

    setObjectivesFormData({
      objectives: objectivesResponseData.objectives,
      caveats: objectivesResponseData.caveats
    });

    setOpenEditDialog(true);
  };

  const handleDialogEditSave = async (values: IProjectObjectivesForm) => {
    const projectData = {
      objectives: { ...values, revision_count: objectivesDataForUpdate.revision_count }
    };

    try {
      await restorationTrackerApi.project.updateProject(id, projectData);
    } catch (error) {
      const apiError = error as APIError;
      showErrorDialog({ dialogText: apiError.message, open: true });
      return;
    } finally {
      setOpenEditDialog(false);
    }

    props.refresh();
  };

  return (
    <>
      <EditDialog
        dialogTitle={EditObjectivesI18N.editTitle}
        open={openEditDialog}
        component={{
          element: <ProjectStepComponents component="ProjectObjectives" codes={codes} />,
          initialValues: objectivesFormData,
          validationSchema: ProjectObjectivesFormYupSchema
        }}
        onCancel={() => setOpenEditDialog(false)}
        onSave={handleDialogEditSave}
      />
      <Box>
        <H3ButtonToolbar
          label="Objectives"
          buttonLabel="Edit"
          buttonTitle="Edit Project Objectives"
          buttonStartIcon={<Icon path={mdiPencilOutline} size={0.875} />}
          buttonOnClick={() => handleDialogEditOpen()}
          toolbarProps={{ disableGutters: true }}
        />
        <Divider></Divider>
        <Box mt={2}>
          <ReadMoreField text={objectives.objectives} maxCharLength={850} />
          {objectives.caveats && <ReadMoreField text={objectives.caveats} maxCharLength={850} />}
        </Box>
      </Box>
    </>
  );
};

export default ProjectObjectives;
