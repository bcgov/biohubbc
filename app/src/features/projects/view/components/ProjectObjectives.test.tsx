import { cleanup, render, fireEvent, getAllByText, waitFor } from '@testing-library/react';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import { UPDATE_GET_ENTITIES } from 'interfaces/useProjectApi.interface';
import { getProjectForViewResponse } from 'test-helpers/project-helpers';
import React from 'react';
import ProjectObjectives from './ProjectObjectives';
import { codes } from 'test-helpers/code-helpers';
import { DialogContextProvider } from 'contexts/dialogContext';

jest.mock('../../../../hooks/useRestorationTrackerApi');
const mockuseRestorationTrackerApi = {
  project: {
    getProjectForUpdate: jest.fn<Promise<object>, []>(),
    updateProject: jest.fn()
  }
};

const mockRestorationTrackerApi = ((useRestorationTrackerApi as unknown) as jest.Mock<
  typeof mockuseRestorationTrackerApi
>).mockReturnValue(mockuseRestorationTrackerApi);

const mockRefresh = jest.fn();

const renderContainer = () => {
  return render(
    <DialogContextProvider>
      <ProjectObjectives projectForViewData={getProjectForViewResponse} codes={codes} refresh={mockRefresh} />
    </DialogContextProvider>
  );
};

describe('ProjectObjectives', () => {
  beforeEach(() => {
    // clear mocks before each test
    mockRestorationTrackerApi().project.getProjectForUpdate.mockClear();
    mockRestorationTrackerApi().project.updateProject.mockClear();
  });

  afterEach(() => {
    cleanup();
  });

  const longData =
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean' +
    ' commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient ' +
    'montes, nascetur ridiculus mus.\n Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.\n' +
    ' Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget,' +
    ' arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo.\n\n' +
    'Nullam dictum felis eu pede mollis pretium. Integer tincidunt. \n Cras dapibus. Vivamus elementum\n' +
    ' semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim.\n\n ' +
    'Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius\n\n ' +
    'laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies\n ' +
    'nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper liber\n\n,' +
    'sit amet adipiscing sem neque sed ipsum. N\n\n';

  it('renders correctly when objectives length is <= 850 characters', () => {
    const { asFragment } = renderContainer();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when objectives length is > 850 characters and caveats is empty', () => {
    const { asFragment } = render(
      <ProjectObjectives
        projectForViewData={{
          ...getProjectForViewResponse,
          objectives: { ...getProjectForViewResponse.objectives, objectives: longData, caveats: '' }
        }}
        codes={codes}
        refresh={mockRefresh}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when both objectives and caveats have length is > 850 characters and are in multiple paragraphs', () => {
    const { asFragment } = render(
      <ProjectObjectives
        projectForViewData={{
          ...getProjectForViewResponse,
          objectives: { ...getProjectForViewResponse.objectives, objectives: longData, caveats: longData }
        }}
        codes={codes}
        refresh={mockRefresh}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when objectives and caveats are < 850 characters and in multiple paragraphs', () => {
    const multilineObjectives = 'Paragraph1\nParagraph2\n\nParagraph3';
    const multilineCaveats = 'Paragraph1\nParagraph2\n\nParagraph3';

    const { asFragment } = render(
      <ProjectObjectives
        projectForViewData={{
          ...getProjectForViewResponse,
          objectives: {
            ...getProjectForViewResponse.objectives,
            objectives: multilineObjectives,
            caveats: multilineCaveats
          }
        }}
        codes={codes}
        refresh={mockRefresh}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('functions as expected with the read more and read less buttons', () => {
    const { container } = render(
      <ProjectObjectives
        projectForViewData={{
          ...getProjectForViewResponse,
          objectives: { ...getProjectForViewResponse.objectives, objectives: longData, caveats: longData }
        }}
        codes={codes}
        refresh={mockRefresh}
      />
    );

    // for finding 'project objectives'
    //@ts-ignore
    expect(getAllByText(container, 'Read More')[0]).toBeInTheDocument();

    //@ts-ignore
    fireEvent.click(getAllByText(container, 'Read More')[0]);

    //@ts-ignore
    expect(getAllByText(container, 'Read Less')[0]).toBeInTheDocument();

    //@ts-ignore
    fireEvent.click(getAllByText(container, 'Read Less')[0]);

    //@ts-ignore
    expect(getAllByText(container, 'Read More')[0]).toBeInTheDocument();

    // for finding 'project caveats'
    //@ts-ignore
    expect(getAllByText(container, 'Read More')[1]).toBeInTheDocument();

    //@ts-ignore
    fireEvent.click(getAllByText(container, 'Read More')[1]);

    //@ts-ignore
    expect(getAllByText(container, 'Read Less')[0]).toBeInTheDocument();

    //@ts-ignore
    fireEvent.click(getAllByText(container, 'Read Less')[0]);

    //@ts-ignore
    expect(getAllByText(container, 'Read More')[1]).toBeInTheDocument();
  });

  it('editing the project objectives works in the dialog', async () => {
    mockRestorationTrackerApi().project.getProjectForUpdate.mockResolvedValue({
      objectives: {
        objectives: 'initial objectives',
        caveats: 'initial caveats',
        revision_count: 0
      }
    });

    const { getByText, queryByText } = renderContainer();

    await waitFor(() => {
      expect(getByText('Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Edit'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.getProjectForUpdate).toBeCalledWith(getProjectForViewResponse.id, [
        UPDATE_GET_ENTITIES.objectives
      ]);
    });

    await waitFor(() => {
      expect(getByText('Edit Project Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Cancel'));

    await waitFor(() => {
      expect(queryByText('Edit Project Objectives')).not.toBeInTheDocument();
    });

    fireEvent.click(getByText('Edit'));

    await waitFor(() => {
      expect(getByText('Edit Project Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Save Changes'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.updateProject).toHaveBeenCalledTimes(1);
      expect(mockRestorationTrackerApi().project.updateProject).toBeCalledWith(getProjectForViewResponse.id, {
        objectives: {
          objectives: 'initial objectives',
          caveats: 'initial caveats',
          revision_count: 0
        }
      });
      expect(mockRefresh).toBeCalledTimes(1);
    });
  });

  it('displays an error dialog when fetching the update data fails', async () => {
    mockRestorationTrackerApi().project.getProjectForUpdate.mockResolvedValue({
      objectives: undefined
    });

    const { getByText, queryByText } = renderContainer();

    await waitFor(() => {
      expect(getByText('Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Edit'));

    await waitFor(() => {
      expect(getByText('Error Editing Project Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Ok'));

    await waitFor(() => {
      expect(queryByText('Error Editing Project Objectives')).not.toBeInTheDocument();
    });
  });

  it('shows error dialog with API error message when getting objectives data for update fails', async () => {
    mockRestorationTrackerApi().project.getProjectForUpdate = jest.fn(() =>
      Promise.reject(new Error('API Error is Here'))
    );

    const { getByText, queryByText } = renderContainer();

    await waitFor(() => {
      expect(getByText('Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Edit'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeInTheDocument();
    });

    fireEvent.click(getByText('Ok'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeNull();
    });
  });

  it('shows error dialog with API error message when updating objectives data fails', async () => {
    mockRestorationTrackerApi().project.getProjectForUpdate.mockResolvedValue({
      objectives: {
        objectives: 'initial objectives',
        caveats: 'initial caveats',
        revision_count: 0
      }
    });
    mockRestorationTrackerApi().project.updateProject = jest.fn(() => Promise.reject(new Error('API Error is Here')));

    const { getByText, queryByText, getAllByRole } = renderContainer();

    await waitFor(() => {
      expect(getByText('Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Edit'));

    await waitFor(() => {
      expect(mockRestorationTrackerApi().project.getProjectForUpdate).toBeCalledWith(getProjectForViewResponse.id, [
        UPDATE_GET_ENTITIES.objectives
      ]);
    });

    await waitFor(() => {
      expect(getByText('Edit Project Objectives')).toBeVisible();
    });

    fireEvent.click(getByText('Save Changes'));

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeInTheDocument();
    });

    // Get the backdrop, then get the firstChild because this is where the event listener is attached
    //@ts-ignore
    fireEvent.click(getAllByRole('presentation')[0].firstChild);

    await waitFor(() => {
      expect(queryByText('API Error is Here')).toBeNull();
    });
  });
});
