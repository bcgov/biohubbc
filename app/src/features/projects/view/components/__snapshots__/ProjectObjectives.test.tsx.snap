// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectObjectives renders correctly when both objectives and caveats have length is > 850 characters and are in multiple paragraphs 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-11"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular"
      style="justify-content: space-between;"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-colorInherit"
      >
        Objectives
      </h3>
      <div
        class="MuiBox-root MuiBox-root-12"
      >
        <button
          aria-label="Edit Project Objectives"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sectionHeaderButton MuiButton-textPrimary"
          data-testid="h3-button-toolbar-Edit"
          id="h3-button-toolbar-Edit"
          tabindex="0"
          title="Edit Project Objectives"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                role="presentation"
                style="width: 1.3125rem; height: 1.3125rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                  style="fill: currentColor;"
                />
              </svg>
            </span>
            <strong>
              Edit
            </strong>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-13"
    >
      <div
        class="MuiBox-root MuiBox-root-14"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Nullam dictum felis eu pede mollis pretium. Integer tincidunt. 
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Cras dapibus. Vivamus elementum
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           nisi.
        </p>
        <div
          class="MuiBox-root MuiBox-root-15"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Read More
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-16"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Nullam dictum felis eu pede mollis pretium. Integer tincidunt. 
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Cras dapibus. Vivamus elementum
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           nisi.
        </p>
        <div
          class="MuiBox-root MuiBox-root-17"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Read More
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProjectObjectives renders correctly when objectives and caveats are < 850 characters and in multiple paragraphs 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-18"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular"
      style="justify-content: space-between;"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-colorInherit"
      >
        Objectives
      </h3>
      <div
        class="MuiBox-root MuiBox-root-19"
      >
        <button
          aria-label="Edit Project Objectives"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sectionHeaderButton MuiButton-textPrimary"
          data-testid="h3-button-toolbar-Edit"
          id="h3-button-toolbar-Edit"
          tabindex="0"
          title="Edit Project Objectives"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                role="presentation"
                style="width: 1.3125rem; height: 1.3125rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                  style="fill: currentColor;"
                />
              </svg>
            </span>
            <strong>
              Edit
            </strong>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-20"
    >
      <div
        class="MuiBox-root MuiBox-root-21"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph2
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph3
        </p>
      </div>
      <div
        class="MuiBox-root MuiBox-root-22"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph2
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Paragraph3
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProjectObjectives renders correctly when objectives length is <= 850 characters 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular"
      style="justify-content: space-between;"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-colorInherit"
      >
        Objectives
      </h3>
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <button
          aria-label="Edit Project Objectives"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sectionHeaderButton MuiButton-textPrimary"
          data-testid="h3-button-toolbar-Edit"
          id="h3-button-toolbar-Edit"
          tabindex="0"
          title="Edit Project Objectives"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                role="presentation"
                style="width: 1.3125rem; height: 1.3125rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                  style="fill: currentColor;"
                />
              </svg>
            </span>
            <strong>
              Edit
            </strong>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div
        class="MuiBox-root MuiBox-root-4"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Et ad et in culpa si
        </p>
      </div>
      <div
        class="MuiBox-root MuiBox-root-5"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          sjwer bds
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProjectObjectives renders correctly when objectives length is > 850 characters and caveats is empty 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-6"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular"
      style="justify-content: space-between;"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 MuiTypography-colorInherit"
      >
        Objectives
      </h3>
      <div
        class="MuiBox-root MuiBox-root-7"
      >
        <button
          aria-label="Edit Project Objectives"
          class="MuiButtonBase-root MuiButton-root MuiButton-text sectionHeaderButton MuiButton-textPrimary"
          data-testid="h3-button-toolbar-Edit"
          id="h3-button-toolbar-Edit"
          tabindex="0"
          title="Edit Project Objectives"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                role="presentation"
                style="width: 1.3125rem; height: 1.3125rem;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                  style="fill: currentColor;"
                />
              </svg>
            </span>
            <strong>
              Edit
            </strong>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-8"
    >
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Nullam dictum felis eu pede mollis pretium. Integer tincidunt. 
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Cras dapibus. Vivamus elementum
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim.
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius
        </p>
        <p />
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
           nisi.
        </p>
        <div
          class="MuiBox-root MuiBox-root-10"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Read More
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
